### PR TITLE
version 1.0.0 - YAML support, new and stable Rule API

### LINKED ISSUE #33: Stabilize and document API for Rules
Hi, I followed the existing rules in the codebase and created a rather convoluted one for my company's requirements. It's working quite well but I'd like to have an API that is documented and reasonably stable and some unit testing in place.

Is this a direction you're interested in for this project?

### LINKED ISSUE #53: Can this project support custom sorting？
Now the artifacts sort by path:
`artifacts = sorted(artifacts, key=lambda x: x["path"])`

I wonder if We can support custom sorting rules, such as:
`artifacts = sorted(artifacts, key=lambda x: (x['path'], x['created']))`

For example, the artifacts `log4j-web` sort by path as follows, 2.6.2 is at the end, but I want to let 2.6.2 at the first when I execute cleanup.
So I think sort by path and  created will slove the problem.

image

### LINKED ISSUE #54: Proposal: YAML format for rules
Engineers who support Artifactory sometimes don't know python but almost everyone is familiar with YAML.

Right now artifactory-cleanup import rules only from YAML, but I think the library is used limited because it requires a bit of knowing how to write python script and be familiar with its syntax.

For me, if I find a useful tool in different language and it's configurable via YAML \ JSON \ TOML - it's fine for me, I use it.
But if I find a tool that requires me to write a few lines of Ruby code to configure it - I'll think twice and try to find another solution.

The basic idea why we used python at the very first step - it gives a lot of comfort for people who know python - IDE suggestions, highlight and flexible mechanism to import from different files 

I think we should move our configuration to YAML format, as follow

------------
**artifactory-cleanup.yaml** (can be changed in CLI)
```yaml
artifactory-cleanup:
  server: https://repo.example.com/artifactory
  # $VAR is auto populated from environment variables
  user: $ARTIFACTORY_USERNAME
  password: $ARTIFACTORY_PASSWORD

  policies:
    - name: Remove all .tmp files older than 7 days
      rules:
        - rule: RepoByMask
          mask: "*.tmp"
        - rule: DeleteOlderThan
          days: 7

    - name: My Docker Cleanup Policies
      rules:
        - rule: RepoByMask
          mask: "docker-*-tmp"
        - rule: DeleteOlderThan
          days: 7

    - name: My Docker Cleanup Policies
      rules:
        - rule: RepoByMask
          mask: "docker-*-tmp"
        - rule: ExcludeDockerImages
          masks:
            - '*:latest'
            - '*:release*'
        - rule: DeleteOlderThan
          days: 7
```

### LINKED ISSUE #60: The behaviour of KeepLatestNVersionImagesByProperty is unclear and possibly incorrect
I want to use`KeepLatestNVersionImagesByProperty` as a safeguard to ensure that a few versions are always retained. I've been debugging the method and I'm still not sure what the intended behaviour is.

Assume the input is (with importance to the ordering):
```
result_artifact = [
    {"properties":{"docker.manifest":"0.1.100"},"path":"foobar/0.1.100"}, # refer to below as A
    {"properties":{"docker.manifest":"0.1.200"},"path":"foobar/0.1.200"}, # refer to below as B
    {"properties":{"docker.manifest":"0.1.99"},"path":"foobar/0.1.99"}    # refer to below as C
]
```
and that the rule has been called like so:
```
KeepLatestNVersionImagesByProperty(count=2, custom_regexp='(^ \d*\.\d*\.\d*.\d+$) ', number_of_digits_in_version=1)
```

The final result ends up being that nothing is deleted. I would have expected that A and B would have been filtered out and that C remains for potential deletion.

If I then alter this line (https://github.com/devopshq/artifactory-cleanup/blob/master/artifactory_cleanup/rules/docker.py#L183) from:
```
key = artifact["path"] + "/" + version_splitted[0]
```
to
```
key = artifact["path"].split("/")[0] + "/" + version_splitted[0]
```
then B and C are filtered out and A remains for potential deletion. This is because the three image versions are able to be grouped together. Previously, the version would be part of the grouping which seems incorrect to me. This still doesn't meet what I expect the result to be.

From here, it seems like the sorting isn't right (assuming what I said above is correct). https://github.com/devopshq/artifactory-cleanup/blob/master/artifactory_cleanup/rules/docker.py#L189
```
key=lambda x: [int(x) for x in x[0].split(".")]
```

### LINKED ISSUE #68: Multiple paths not working
Hi

I was trying to write some cleanup policies that make use of the multiple path feature, but it does not seem to work with the latest Artifactory version

Artifactory version :  7.41.12 rev 74112900
Artifactory-cleanup version : 0.4.2

Be aware that I ofuscated repo, microservice and url of our Artifactory

Cleanup Policies

```python
from artifactory_cleanup import rules
from artifactory_cleanup import CleanupPolicy
RULES = [
  CleanupPolicy(
  'Test',
  rules.Repo('npm-release-local'),
  rules.IncludePath(['*@scope/microservice*', '*microservice*']),
  rules.IncludeFilename('*-test*'),
  rules.KeepLatestNFiles(count=5),
  ),
]
```

Output

```
********************************************************************************
Verbose MODE
Simulating cleanup actions that will occur on 2022-09-13
Add AQL Filter - rule: Repo - Apply the rule to one repository.
Get from npm-release-local
Checking the existence of the npm-release-local repository
The npm-release-local repository exists
Add AQL Filter - rule: IncludePath - Apply to artifacts by path / mask.
Add AQL Filter - rule: IncludeFilename - Apply to artifacts by name/mask.
Add AQL Filter - rule: KeepLatestNFiles - Leaves the last (by creation time) files in the amount of N pieces. WITHOUT accounting subfolders
********************************************************************************
AQL Query:
Add AQL Text - rule: Repo - Apply the rule to one repository.
Add AQL Text - rule: IncludePath - Apply to artifacts by path / mask.
Add AQL Text - rule: IncludeFilename - Apply to artifacts by name/mask.
Add AQL Text - rule: KeepLatestNFiles - Leaves the last (by creation time) files in the amount of N pieces. WITHOUT accounting subfolders
Before AQL text: items.find({"$and": [{"repo": {"$eq": "npm-release-local"}}, {"path": {"$match": ["*@scope/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat")
After AQL text: items.find({"$and": [{"repo": {"$eq": "npm-release-local"}}, {"path": {"$match": ["*@scope/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat").sort({"$asc" : ["created"]})

items.find({"$and": [{"repo": {"$eq": "npm-release-local"}}, {"path": {"$match": ["*@scope/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat").sort({"$asc" : ["created"]})
********************************************************************************
Add AQL Text - rule: Repo - Apply the rule to one repository.
Add AQL Text - rule: IncludePath - Apply to artifacts by path / mask.
Add AQL Text - rule: IncludeFilename - Apply to artifacts by name/mask.
Add AQL Text - rule: KeepLatestNFiles - Leaves the last (by creation time) files in the amount of N pieces. WITHOUT accounting subfolders
Before AQL text: items.find({"$and": [{"repo": {"$eq": "npm-release-local"}}, {"path": {"$match": ["*@scope/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat")
After AQL text: items.find({"$and": [{"repo": {"$eq": "npm-release-local"}}, {"path": {"$match": ["*@scope/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat").sort({"$asc" : ["created"]})

Traceback (most recent call last):
  File "/home/alex/.local/bin/artifactory-cleanup", line 8, in <module>
    sys.exit(ArtifactoryCleanupCLI())
  File "/home/alex/.local/lib/python3.10/site-packages/plumbum/cli/application.py", line 178, in __new__
    return cls.run()
  File "/home/alex/.local/lib/python3.10/site-packages/plumbum/cli/application.py", line 624, in run
    retcode = inst.main(*tailargs)
  File "/home/alex/.local/lib/python3.10/site-packages/artifactory_cleanup/cli.py", line 114, in main
    for summary in cleanup.cleanup(
  File "/home/alex/.local/lib/python3.10/site-packages/artifactory_cleanup/artifactorycleanup.py", line 54, in cleanup
    artifacts = policy.get_artifacts()
  File "/home/alex/.local/lib/python3.10/site-packages/artifactory_cleanup/rules/base.py", line 214, in get_artifacts
    r.raise_for_status()
  File "/usr/lib/python3.10/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error:  for url: https://selfhosted.artifactory.com/artifactory/api/search/aql

```

And taking a look into our Artifactory logs we get this

```
2022-09-13T16:49:56.667Z [jfrt ] [ERROR] [247cfbc55f7e33d4] [o.a.r.r.a.AqlResource:101     ] [ttp-nio-8081-exec-26] - Fail to execute query: items.find({"$and": [{"repo": {"$eq": "npm-release-local"}}, 
{"path": {"$match": ["*@scoped/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat").sort({"$asc" : ["created"]}): 

Failed to parse query: items.find({"$and": [{"repo": {"$eq": "npm-release-local"}}, {"path": {"$match": ["*@scoped/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat").sort({"$asc" : ["created"]}), 
it looks like there is syntax error near the following sub-query: ["*@scoped/microservice*", "*microservice*"]}}, {"name": {"$match": "*-test*"}}]}).include("*", "property", "stat").sort({"$asc" : ["created"]})
```

We get the same error trying to use multiple items in `IncludeFilename` , if we write two separated cleanup policies they work just fine 

Thanks
Alex

### PR DESCRIPTION
**Pull request by allburov**
- fix #33 
- fix #54
- fix #68
- fix #53
- fix #60
- Support JCR #59

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,7 +33,7 @@
 ## 0.4.0
 
 - Remove `--remove-empty-folder` option. Artifactory provides corresponding built-in functionality already
-- Change the `delete_empty_folder` rule to not depend on an external plugin, but directly delete files from this script
+- Change the `delete_empty_folders` rule to not depend on an external plugin, but directly delete files from this script
```
Comment (allburov) [inline CHANGELOG.md:36]: Change it back, it's a history record

**Comment by allburov** [comment]
@Pro @zhan9san Hi!
Could you have a look at the changes for YAML configuration? 
Readable readme is here: https://github.com/devopshq/artifactory-cleanup/tree/yaml-configuration

If you have a time - could you run the new version against your configuration files one time to make sure it works the same way it used to work.
```bash
python3 -mpip install git+https://github.com/devopshq/artifactory-cleanup.git@yaml-configuration
```

Suggestions and notes are welcome!

**Comment by donhui** [comment]
@allburov 
I have tested cleanup docker images, it works for me.

**Comment by Pro** [comment]
Will not have time to test it. Just a quick note from my side:

Since this is a new major release, you should use the chance to get rid of old/deprecated stuff.
IMHO backward compatibility does not need to be there for major version bumps.

**Comment by allburov** [comment]
@donhui thank you man!
@zhan9san sure, waiting for your feedback!

@Pro good note! 
Do you mean to remove these features or I missed something? 
1. Load rules from python file
2. CLI parameters artifactory-server, user, password
3. `under_score` names for Rules

**Comment by Pro** [comment]
> @donhui thank you man! @zhan9san sure, waiting for your feedback!
> 
> @Pro good point! Do you mean to remove these features or I missed something?
> 
> 1. Load rules from python file
> 2. CLI parameters artifactory-server, user, password
> 3. `under_score` names for Rules

Mainly thought about removing old/deprecated features, like the ones you already mention. Did not look in detail at all the changes.

**Comment by allburov** [comment]
@Pro great, I agree with you, it was so annoying  to keep them in mind when changing code... I'll remove all old features!

**Comment by donhui** [comment]
@allburov
I have a suggestion.
When deleting an artifact, can we print the size of each artifact to be deleted into the log?
Or, can the logs printed when deleting artifacts be expanded? So that users can customize the log.
Thanks.

**Comment by allburov** [comment]
@donhui 
>can we print the size of each artifact

Added it! 

>deleting artifacts be expanded?

I was thinking about adding API for cleanuppolicy too - to handle some special case like cleaning docker or conan repositories. But for now it's too rare cases, so we can handle them in configuration just by using current rules.

**Pull request merged**